### ISSUE TITLE
Investigation: non-BOSH worker operation lifecycle

### ISSUE THREAD

**Issue by topherbullock** [concourse/concourse#1457]
## Challenge

Operators of concourse clusters (non-BOSH since it has drain scripts to handle worker lifecycle) often run into issues when operating a cluster of workers (creating, restarting, scaling, etc). 

## A Modest Proposal
Look into the following operation setups and document the pain points around operating (creating, recreating, scaling) worker clusters:
- Binaries on VMs
- `docker-compose`'d
- Kubernetes pods (via Helm, or manually deployed)

If there is a sane workflow that involves the tools available currently (`concourse land-worker`, `concourse retire-worker`, `fly prune-worker` ), we should improve the documentation around how these commands should be used to operate a cluster where the workers undergo regular recreation, scaling, etc.

If there are gaps in operating these scenarios (workers can't be recreated easily using the tools available, errors happen due to data persistence on 'normal' deployments of K8s, docker compose, binaries ) we should strategize and discuss how to tackle them.

**Comment by JohannesRudolph** [comment]
Please also see my comment here for a reproduction/scenario where the current worker lifecycle is a problem: 
https://github.com/concourse/concourse/issues/844#issuecomment-300175241

Additionally, I have noticed that `fly prune-worker` does not provide immediate relief when killing of old workers that won't come back. Sometimes our two node cluster takes more than an hour after all old workers were pruned until all build pipelines show no more errors, e.g. with check containers. 

I'm curious about concourse retire-worker, never heard of that. Is this something that the concourse binary could trigger on receiving a signal from `docker stop` or `docker kill`?

**Comment by william-tran** [comment]
@topherbullock I've resigned to having workers register with unique names per invocation: https://github.com/kubernetes/charts/compare/master...autonomic-ai:concourse-worker-lifecycle-management, but this can result in stalled workers if processes don't gracefully shutdown (eg exec into container and kill -9), and containers on stalled workers aren't rescheduled onto running workers. In a test I've needed to prune the stalled workers before the new worker would start getting containers. Are there any plans to auto-remove workers that are stalled for a period of time?

**Comment by william-tran** [comment]
Fixed the issue above with stalled workers by keeping the worker names consistent across invocations. A liveness probe that watches the logs will restart a stuck worker if it detects FS panic or `unknown volume`, and a preStop hook takes care of retirement. Lets hope this suffices.

**Comment by topherbullock** [comment]
@william-tran some clarifications on worker names, and state as it exists now:

If the hosts are the same - in that the persistent disk ( or whatever storage volume they're using ) for Garden and BaggageClaim are the same disk with existing containers and volumes hanging around - then the workers should register with the same name. These workers should _ideally_ be brought down using the `concourse land-worker` command when an operator knows they will be offline temporarily, but come back up with the same persistent storage.

If the hosts can be considered unique, and there is no persistent data left around on them, then registering with a unique name is the way to go. These workers should _ideally_ be brought down using the `concourse retire-worker` command when an operator knows they will be retired permanently, and **not** come back up with the same persistent storage.

There's definitely some issues around `STALLED` workers still keeping their containers and those containers not being rebalanced across the whole cluster. There's [an existing issue](https://github.com/concourse/concourse/issues/1458) about the error messages resulting from this, but we should also tackle the other issue you've mentioned of rebalancing after some time.

**Comment by william-tran** [comment]
@topherbullock Its seems that when `concourse worker` starts on k8s, the contents of the /concourse-work-dir folders get wiped out, eg `depot` and `volumes/live`. This is the case even when using  Persistent Volumes. I've confirmed this overriding the entrypoint and adding a sleep before `concourse worker`, then kill -9 ing the worker process to trigger a container restart. The contents of `/concourse-work-dir` survive after container restart, but then get wiped out when `concourse` starts. I see in `concourse worker --help`

```
--garden-destroy-containers-on-startup             Clean up all the existing containers on startup. [$CONCOURSE_GARDEN_DESTROY_CONTAINERS_ON_STARTUP]
```

This is unset, what's the default behaviour?

**Comment by william-tran** [comment]
On the above behaviour, I think https://github.com/concourse/concourse/issues/1255 needs to be re-opened.

**Comment by topherbullock** [comment]
#1458 captures fixing the root cause for those errors. That one was closed off in favor of a larger-scoped fix for how we react to resources existing on workers which are in a stalled state.

**Comment by topherbullock** [comment]
> This is unset, what's the default behaviour?
Default is false, and in that case Garden will try and restore from the depot dir.

**Comment by WillemMali** [comment]
I'm currently integrating Concourse with my k8s/helm setup and I'm planning on upstreaming it to the main Helm chart repo, if anyone wants it I can upload what I have right now.

**Comment by william-tran** [comment]
@WillemMali have you seen what's there right now? https://github.com/kubernetes/charts/tree/master/stable/concourse and my PR to it for improving worker lifecycle management https://github.com/kubernetes/charts/pull/2109

**Comment by WillemMali** [comment]
I have, it uses StatefulSets which I'm unfamiliar with (and Helm doesn't support them), so I'm making a chart that works with Deployments (and which has a few footguns removed, like all default credentials).
-------- Original message --------From: Will Tran <notifications@github.com> Date: 9/22/17  16:43  (GMT+01:00) To: concourse/concourse <concourse@noreply.github.com> Cc: WillemMali <willem.mali@gmail.com>, Mention <mention@noreply.github.com> Subject: Re: [concourse/concourse] Investigation: non-BOSH worker operation
  lifecycle (#1457) 
@WillemMali have you seen what's there right now? https://github.com/kubernetes/charts/tree/master/stable/concourse and my PR to it for improving worker lifecycle management kubernetes/charts#2109

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/concourse/concourse","title":"concourse/concourse","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/concourse/concourse"}},"updates":{"snippets":[{"icon":"PERSON","message":"@william-tran in #1457: @WillemMali have you seen what's there right now? https://github.com/kubernetes/charts/tree/master/stable/concourse and my PR to it for improving worker lifecycle management https://github.com/kubernetes/charts/pull/2109"}],"action":{"name":"View Issue","url":"https://github.com/concourse/concourse/issues/1457#issuecomment-331466982"}}}

**Comment by william-tran** [comment]
@WillemMali best to move your reasons for changes to the chart to an issue against https://github.com/kubernetes/charts

**Comment by WillemMali** [comment]
I was a bit rude there, sorry. Thank you for your suggestion.

**Comment by EugenMayer** [comment]
Still an issue here, can be reproduced using `docker compose up` and then either `down/up`  or `stop/start` with only a single worker setup like here https://github.com/EugenMayer/concourseci-server-boilerplate

It seems like after restarting the stack ( including the worker ) the worker is stalled.

```
fly -t lite workers
name          containers  platform  tags  team  state    version
ef560177674d  1           linux     none  none  running  1.2    


the following workers have not checked in recently:
```

Interestingly, when doing `down/up` which implies that all anon-volumes are deleted, the stack-self heals after about 5 minutes ( timeout of something )? with `stop/start` which would also persist the anon volumes, the issues does never go away. 

That issue can be easily reproduced every single time on https://github.com/EugenMayer/concourseci-server-boilerplate

```
git clone https://github.com/EugenMayer/concourseci-server-boilerplate
cd concourseci-server-boilerplate
# adjust CONCOURSE_EXTERNAL_URL to your docker-machine ip (see readme)
docker-compose up -d
# this deploys the example example pipeline and runs the job
# you probably need to adjust your docker-machine ip for the login
./run_vault_test.sh
docker-compose stop
docker-compose start
fly -t lite login -c http://127.0.0.1:8080 -u concourse -p changeme
fly -t lite trigger-job -j main/vault-test -w
```

**Related, but different issue**
Pruning the work spawns a new registration right away ( it seems ) but the job will fail since resources seem to no longer get pulled properly:
![image](https://user-images.githubusercontent.com/136934/32893596-6d50950a-cada-11e7-87b1-24d52ff0bf6c.png)
the pipeline is public and very simple https://github.com/EugenMayer/concourseci-server-boilerplate/blob/master/examples/vault-based/pipeline.yml and works when freshly deployed.

Destroying the pipeline and configuring it again does fix the resource issue. I guess that the actual new worker is to be expected to have some cache present of the resource, he might actually not have ( the new one ) .. destroying the pipeline and redeploying it probably invalidates that cache expectation

**Comment by EugenMayer** [comment]
the issue comes down to the lost worker-state, but also to the changing hostname. You can fix this by 
 - https://github.com/EugenMayer/concourseci-server-boilerplate/blob/master/docker-compose.yml#L51
 - https://github.com/EugenMayer/concourseci-server-boilerplate/blob/master/docker-compose.yml#L58

This will ensure the worker gets properly registered on stack restarts and upgrades. 

Still that wont fix the issues overall, since after the stack comes back up, you will get different other issues: 
![image](https://user-images.githubusercontent.com/136934/32895504-945af3e2-cae0-11e7-8efc-0b6ac2501ccc.png)

while the job is stalling in `preparing build` This changes after 1 minute, then the recource is fixed but the actuall job cannot fetch the task or in other words, the resources seem to not be properly pulled / placed on the wrong volume / assumed to be still cached / whatever.

![image](https://user-images.githubusercontent.com/136934/32895617-ef230de6-cae0-11e7-8a62-527af2cba216.png)

Again, destroying the pipeline and readding it, will fix this. I could never recover this pipeline, no matter what i did.

I assume there is even more state which is not properly persisted but assumed to be persistent ( thats why it works on BOSH )..which is not in /worker-state .
I think its fine to assume that /worker-state should be a anon-volume, so not fully persistent, but then the workers registration should be gone too, which is most probably part of the DB, which is fully persisted. This inconsistency leads to the first issue, the latter one - i do not know.

The issue with the hostname, well its a conceptual flaw - you never fixate on the hostname with elastic concepts in docker - concourse either needs a dynamic service registration itself or use consul or similar to dynamically register agents or discover them. Setting the hostname to `worker1` `worker2` is a workaround, but its not fixing the actual issue

**Comment by EugenMayer** [comment]
wrapping my head around this a bit more i guess the reason for the latter issue is the docker-in-docker engine with its volumes.

When a the worker is stopped using `down/stop` those volumes are lost, since the docker-state is not persisted at all ( not even a anon volume ). this should be the reason for the issues with the resources

**Comment by mbjelac** [comment]
@EugenMayer  - I more or less reported the same thing, except I reported the related issue in its own github issue:
#1796 
#1797 

The only difference is that you invested more effort in reproducing it in detail. Thanks!

**Comment by topherbullock** [comment]
@EugenMayer Yeah, the `docker-compose stop` issue is a known problem 👍  The intent of this issue is to solve these problems or document how users can better account for how worker pools are managed. 

Workers can be managed using `concourse land-worker`, `concourse retire-worker`, and `fly prune-worker` commands, but these are manual decisions an operator needs to make, knowing the details of how the hosts running the workers are managed.

Here's the ideal flow for stopping a worker that is going down temporarily (coming back with the same name and disk):  
- use `concourse land-worker` to stop new work from being scheduled 
- do whatever operations-y things for bringing the host down
- bring the host back up; it will re-register with the worker pool and no longer be landed

If the worker isn't going to re-register with the same name and persistent disk, it should be retired using `concourse retire-worker`.

This works fine with BOSH because we have drain scripts which deal with the lifecycle of workers, but the goal of this issue is to document the steps I describe above, or instrument other deployment scenarios to the same capabilities as BOSH.

**Comment by EugenMayer** [comment]
@topherbullock so why dont we use a trap for KILL or similar on the worker to properly land it before stopping it?
Docker usually does not simply kill a process, so the worker should have cleaned up before it exists in the first place - but anyway, a trap could just do what we want here.

i guess `concourse land-worker` are yet cli only commands, no way to trigger this using `fly` right.

I cannot see how land should be a manual decision, AFAICS a worker is broken when it gets restarted every single time, right ( when using docker ). I think there is not really a choice here, is it?

We do not need to have this trap for BOSH if its working there, but for at least docker scenarios it makes sense in the first place, doesnt it

So a wrapper like this should do it
```
function shutdown()
{
    echo "Shutting down Standalone"
    concourse land-worker `hostname`
}

# Allow any signal which would kill a process to stop standalone
trap shutdown  KILL TERM

worker &

wait $!
```

i assume the name is the hostname, so we can derive that argument from that. Anything i missed?

**Comment by topherbullock** [comment]
@EugenMayer That trap works for the case when the worker will definitely come back, and come back with the same state volume, but will leave the worker landed when the container is stopped and will never come back.

Ideally workers running in containers can (and I would argue _should_) be considered ephemeral without any state left on them, so they can be scaled up and down, recreated, or destroyed without persistence. The persisting the Garden state directory to a volume just leaves a lot of junk hanging around from old container handles, and doesn't really account for worker stop/starts as the process associated is very likely killed and the 'container' is just a rootfs in a volume that needs to be cleaned up or it will sit there forever.

To me the preferred Docker / Kubernetes approach shouldn't involve being precious about worker containers, their names, or any state left on them.

Getting back to the end goal if this issue, I'd like to see a flow which accounts for this, rather than treat containers as special hosts with state on them, which is antithetical to Concourse's ideal treatment of workers. I'm also thinking I'd like to improve the worker registration to make the `retire / land / stall` decision more automatic, rather than special casing the logic for specific uses of Docker, the binaries, Kubernetes, and BOSH,

**Comment by EugenMayer** [comment]
@topherbullock i fully agree, that yet we are far away from the optimal case and that solving the issue properly would involve a very different concept - so i am on you on the same page.

Still, this issue should probably go into docker-concourse, since if you offer this http://concourse.ci/docker-repository.html in your docs - in should somehow also work, which is not the case right now. This cannot be used in production at all, and that is what we could fix for the time being, why you maybe come up with a service-registry or whatever internal concept that might be.

Right now, providing that trap-wrapper for worker only in docker-concourse will help every single person running that docker container. Since when you do stop or down, you do not do this because you want to remove a single worker, you want to pause or upgrade your stack - and this always falls into the `worker-land` case.

Anyway whoever this turns out here, i am fine with the decision (using that trap already for me anyway, be it a corner case or not)

**Comment by EugenMayer** [comment]
Just in case anybody needs that, i created https://github.com/EugenMayer/docker-image-concourseci-worker-solid / https://hub.docker.com/r/eugenmayer/concourse-worker-solid ... made a test case with https://github.com/EugenMayer/concourseci-server-boilerplate .. as you see https://github.com/EugenMayer/concourseci-server-boilerplate/blob/master/docker-compose.yml#L51 

With that setup, i can stop/start down/up workers without running into any draining issues anymore. I am using the trap setup above, just always using `retire-worker` and never `land-worker` since the latter is only working partially - you end up having fixed the `file not found` issue, but get a new `handle not found` issue on resources.

With the trap, you do no longer need fixed hostnames and no longer need to persist /workers-state - the first one is not needed since workers are de-registered anyway and registered again. The latter one is useless since `land-worker` is broken anyway.

Using `eugenmayer/concourse-worker-solid:3.6.0` fixed the whole case for me.

**Comment by mbjelac** [comment]
@EugenMayer - I just tried `eugenmayer/concourse-worker-solid:3.6.0`. The web component fails to start up. Here are the last few lines from the console before it dies:

```
+ /worker-state/3.6.0/assets/iptables/sbin/iptables -w -N w--input
+ /worker-state/3.6.0/assets/iptables/sbin/iptables -w -F w--input
iptables v1.4.21: can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.

dns-proxy exited with nil

baggageclaim exited with nil
```

Yes, I tried running it as root. :)

I'm starting it up together with the DB and worker components using `docker-compose` with [this definition](https://gist.github.com/mbjelac/0579875d0c7ff413c14bdabf6d00a0d8).

**Comment by EugenMayer** [comment]
@mbjelac please open an issue in the repo itself https://github.com/EugenMayer/docker-image-concourseci-worker-solid - this issue should not be used for such things. I am happy to help you there, just copy in paste it

**Comment by mbjelac** [comment]
@EugenMayer - yes, sorry. Issue created. (My first github issue with id 1, horaaay! :))

**Comment by EugenMayer** [comment]
@topherbullock if you like, please remove our laste for comments ( including this one ) - sorry for the offtopic stuff @ all

**Comment by topherbullock** [comment]
@pivotal-jwinters and I had some thoughts about this while working on #1027 and discussing having the workers' baggageclaim initialize with a certs volume;

We could have the workers advertise their existing handles for containers and volumes when registering, as well as a unique identifier (generated on process start maybe? persistent somewhere tied to worker state?)  which allows ATC to more reliably determine whether it can be considered the "same" worker or effectively a new worker. 

Advertising the volumes also enables the ATC to cross-reference the state on the worker with its assumed state, and determine whether the ATC has volumes in the db which aren't on the worker's baggageclaim, or vice-versa (tl;dr no more "unknown handle" errors)

**Comment by EugenMayer** [comment]
@topherbullock i try to make up my mind, maybe thats of any use for you:

1. in yet any orchestration in know, hostname resets go with anon-volume resets. That means, if we want to make the worker-state persistent (named volume) which would be required for some of your thoughts, we cannot go with the hostname being the identifier - since we could have the same state (name volume) but new hostname - mixed up

2. I think it might no be needed to "advertise" if we can just keep the state consistent? If a worker shuts down, and the actual docker-engine and volumes / overalyfs is all part of `/worker-state` - it should come up right that same way as it has been shudown. Thus, just ensuring we do not lose any state which is expected by bagageclaim or ATC, will make it consistent out of the box + it ensures we keep as much as possible and restarts become just restarts.

3. I think checking if a volume is referenced in the DB but not present or vise versa is a huge plus for advertising actually, since a lot of people did mention "bleeding" somehow over time. So maybe, even if 2 ensure "maximum state persistance" advertising could help cleaning up a lot better.

Right now, even if i fixate the hostname to a static string and then use a named volume for state it will not work, because the state will not be recovered properly ( even with proper worker-land prior the restart ). Can we find a reason for that?

**Comment by topherbullock** [comment]
Yeah,  @EugenMayer I'm kind of just brain-dumping on this thread to try and collect thoughts.
Regarding: 
1. Moving to a unique identifier other than the worker name would move us from relying on the hostname as the only way to identify the workers.
2 & 3 . The idea behind advertising is to ensure that the desired state in the DB matches the state the worker has currently in its Garden + Baggageclaim. This means we can account for whatever changes have happened (if any) to the state volumes / directory, and helps us be agnostic of how those are actually managed.

**Comment by EugenMayer** [comment]
Advertising sounds great, but maybe rather hard to really get right. Sounds like a far goal to me - any way we could do this in 2 milestones, first with unique IDs + storage check, second advertice to "sanatize further and validate/sync" ?

**Comment by william-tran** [comment]
@topherbullock @EugenMayer and everyone else, I really appreciate the attention you're giving to this issue. 

@EugenMayer I ran into the same issue with land vs retire worker running in k8s with the helm chart, I found retire to be the better option and updated the chart to manage worker lifecycle as a result. See https://github.com/kubernetes/charts/pull/2109/files#diff-33a16a0789cde854c77e65e7774aa2beL35 for the related changes. We're trying to solve the same problem so I want to give you a heads up on my approach. 

@topherbullock Though I'm running into more issues around detecting fatal errors that should trigger a restart. I had `unknown volume` as a fatal error, but once I moved to multiple concourse-web pods, I would get lots of these. Looking at the worker logs, I saw a volume was requested to be deleted and that request being successful, but then another request for deletion would be received (I'm guessing from a different atc) and would log  `unknown volume`, when really the worker could continue taking on work. I've removed `unknown volume` from the list of fatal errors, and am monitoring closely to see if I can catch the truly unrecoverable condition.

In all this, I've seen an opportunity to better clean up before starting the worker under the same name, something like:
```
          rm -rf /concourse-work-dir/*
          while ! concourse retire-worker --name=${HOSTNAME} | grep -q worker-not-found; do
            sleep 5
          done
          concourse worker --name=${HOSTNAME} | tee -a /concourse-work-dir/.liveness_probe
``` 
I'm relying on this log to determine whether the namesake worker is still registered or not:
https://github.com/concourse/tsa/blob/d563b2ac75c8e47e2342baaac205370765f8339b/retirer.go#L53

Sure it's hacky, but what can I do? Do you see any issues arising from calling `retire-worker` multiple times?

**Comment by william-tran** [comment]
Super Simple Solution: expose `delete-worker` to the `concourse` command. Then I can run this before I start the worker to ensure state on both sides is clean. I've tried `curl -X DELETE https://my-concourse.com/api/v1/workers/my-worker-name` and it works wonderfully, except I can't actually do that from the worker.

**Comment by emcniece** [comment]
🔺  I'd like to document a red herring that just cost me 2 days - this issue is where all roads lead for the `file not found` error, please let me know if I should post this elsewhere.

```yml
- get: my-docker-image
  params: {skip_download: true}   # <-- THIS IS A PROBLEM

- task: my-task
  image: my-docker-image
  file: tasks/my-task.yml

- put: my-docker-image
  build: docker/my-docker-image
```

If you flush worker state and rerun a pipeline with a config as defined in this example, the `my-task` step will fail with the `file not found` error because no previous builds of this image have been created/cached in Concourse.

If you are trying to get better performance out of building Docker images in a pipeline, use the [cache or load_base methods instead](https://stackoverflow.com/a/44959191/943540).

This is expected behaviour, but _wow_ I spent a lot of time upgrading and debugging workers before ripping my pipelines apart.

**Comment by william-tran** [comment]
@vito about trying to retire a stalled worker, here's the original issue I opened: https://github.com/concourse/concourse/issues/1919

**Comment by sean-abbott** [comment]
I would just like to put in a plug/vote for workers being able to gracefully land themselves. I.e., I would like to have a command I can give a worker that says "you're shutting down now, let ATC know".

This will make being able to use AWS autoscaling much easier. If I have to run a command on the web server, it makes it hard(er) to have an autoscaling worker pool.

**Comment by leshik** [comment]
I've been hit by this issue today when DigitalOcean restarted the hypervisor with my concourse installation. I'm using `docker-compose` deployment. After the reboot, the worker became stalled and ATC complained:
```
{"timestamp":"1521691858.311057329","source":"atc","message":"atc.container-collector.run.orphaned-containers.queue.worker-not-found","log_level":1,"data":{"session":"54.1078.1.21","worker-name":"5f2c668feaa4"}}
```
So I issued `fly prune-worker` to clean things up. But unfortunately, after that a new worker didn't come up and ATC complained `failed-to-initialize-new-container` with `no workers` error. I had to issue `docker-compose down` and then `docker-compose up` to make it work.

This last thing is especially frustrating, how the need to restart worker's container manually can be avoided?

**Comment by vito** [comment]
Quick update: we've merged and finished up https://github.com/concourse/atc/pull/247 which adds support for ephemeral workers. Ephemeral workers will just immediately go away instead of stalling. This will be off by default, but we're willing to change that for some scenarios (perhaps Docker-Compose).

This is just one piece of the puzzle but it's there now!
